### PR TITLE
 feature(endpoint): create article endpoint [start #45]

### PR DESCRIPTION
**Pull request by beblicarl**
#### Pre-Submission Checklist

-   [x] Your pull request targets the `main` branch.
-   [x] Branch name is descriptive and follows the guidelines [here](https://github.com/andela/bestpractices/wiki/Git-naming-conventions-and-best-practices).
-   [x] All tests pass . Use `git commit --amend` to amend any fixes.

#### Checklist:

-   [x] All tests are passing.
-   [x] Tested changes locally.
-   [x] Improved documentation or readme where necessary.
-   [x] Your commit messages format follows our commit message guideline [here](https://github.com/andela/bestpractices/wiki/Git-naming-conventions-and-best-practices#commits).

#### What does this PR do?

The PR creates the article endpoint
#### How should this be manually tested? 

 
1.Start the server with `npm start`
2. Post sample data bellow to` http://localhost:3000/api/v1/articles `
```
{ 
"id": 6,
"userId" : 1,
"title" : "Rojitech's Frontend Club",
"image" : "Rojitech",
"content" : "dfdgdgsgsdfdfgfgfd",
"published" : "false",
"createdAt" : "{{CurrentDateTime}}"
}
```
3. This should return the response body in the response console of an API client (eg Postman)
 

##### Any background context you want to provide?

This endpoint is used for posting articles to the database
#### What are the relevant issues?
1. Creating an endpoint for posting article data

#### Screenshots (if appropriate)

Screenshot from 2022-08-15 13-22-01

#### Does it require any deployments updates ?

https://github.com/tbf-challenge/teamwork-be/pull/56

### CODE REVIEW

**Comment by jocrah** [comment]
hi @beblicarl, nice to meet you once again. i've invited myself here so lets begin, shall we :sweat_smile:

File: src/routes/users.js [hunk 1]
```diff
@@ -12,19 +14,14 @@ const getUser = async (req, res) => {
...
-router.route('/')
-.get(fetchUsers)
-.post(createUsers)
+router.route('/').get(fetchUsers).post(createUsers)
 
-router.route('/:id')
-.get(getUser)
-.patch(updateUser)
-.delete(deleteUser)
+router.route('/:id').get(getUser).patch(updateUser).delete(deleteUser)
```
Comment (jocrah) [inline src/routes/users.js:25]: curious about this change. does the linter reformat this by default or this is by personal preference?

File: src/routes/users.js [hunk 2]
```diff
@@ -12,19 +14,14 @@ const getUser = async (req, res) => {
 }
 const updateUser = () => {}
 const deleteUser = () => {}
-const checkId = (req, res, next , val ) => {
+const checkId = (req, res, next, val) => {
```
Comment (jocrah) [inline src/routes/users.js:17]: @beblicarl could you shed some light on what this should do?

**Review by jocrah — COMMENTED** [review]
not to get ahead of myself since this is a WIP. Could you shed more light on what is left?

**Comment by beblicarl** [comment]
> hi @beblicarl, nice to meet you once again. i've invited myself here so lets begin, shall we sweat_smile

Heya @jocrah , nice to meet you too, let's begin sweat_smile

File: src/routes/users.js (hunk 1 above)
Comment (beblicarl) [reply src/routes/users.js:25]: the user.js file serves as a template for how the various endpoint should be built

File: src/routes/users.js (hunk 2 above)
Comment (beblicarl) [reply src/routes/users.js:17]: the user.js file serves as the template of how the post.js should be built. I worked on the post.js file @jocrah

File: src/routes/users.js (hunk 1 above)
Comment (jocrah) [reply src/routes/users.js:25]: okay cool but still curious as to the change from the multi-line to one-line. wanted to know if it was reformatted by the linter

Comment (beblicarl) [reply src/routes/users.js:25]: no, it wasn't , it was a personal choice

File: src/routes/users.js (hunk 2 above)
Comment (jocrah) [reply src/routes/users.js:17]: Got it but our pr here has changes in the user.js so trying to confirm some things there as well. just curious to know what the `checkId` does or you mean to say it wasn't added by you?

File: src/routes/users.js (hunk 1 above)
Comment (jocrah) [reply src/routes/users.js:25]: okay cool

**Comment by jocrah** [comment]
@beblicarl anyway was asking that since this is a WIP, i would like to know what you feel is left to do for it to be ready

**Comment by beblicarl** [comment]
> @beblicarl anyway was asking that since this is a WIP, i would like to know what you feel is left to do for it to be ready

some of the properties and endpoint will be renamed according to the names provided in the pdf, this is a draft just to get the endpoints working

File: src/routes/users.js (hunk 2 above)
Comment (beblicarl) [reply src/routes/users.js:17]: It was initially meant to be used to check for authentication , but it is something I will take out in my next commit @jocrah

Comment (jocrah) [reply src/routes/users.js:17]: Okay, I think since we're learning we should go by the ideal approach.

Let's always make sure our changes mainly deal with just what we are mainly working on or the scope of the issue at hand.

Otherwise, it becomes something the reviewer has to review and that becomes a bit confusing as well

So instead of doing additional changes, you can instead reset the file to remove your changes by doing `git checkout origin/main <file_path>` and then commit it.

Kindly let me know if you get me.

Comment (beblicarl) [reply src/routes/users.js:17]: yeah, i get you. You are right, it doesn't have anything to do with the actual issue at hand. I will rest the file to remove my changes

**Comment by jocrah** [comment]
> some of the properties and endpoint will be renamed according to the names provided in the pdf, this is a draft just to get the endpoints working

@beblicarl okay cool, any reason why the properties and endpoint can't be renamed now? or there's more to that aspect?

**Comment by jocrah** [comment]
a few things to disturb you about your pr description :smile: 

- the checkboxes shouldn't have a space else it doesn't show well. It should be `[x]` instead of `[x ]` 

- one of the main things is to help another person test easily so you have to assume they know nothing. 
e.g of such a manual test
1. Start the server `npm run start`
2. Post the sample data below to the endpoint `/api/v1/articles`
```json
{
  "name": "Yes",
  "message": "No"
}
```
3. It should successfully return `so so and so` response
- Screenshots will not be too necessary in that case
- Also link the relevant issue if possible

**Comment by beblicarl** [comment]
> a few things to disturb you about your pr description smile
> 
> * the checkboxes shouldn't have a space else it doesn't show well. It should be `[x]` instead of `[x ]`
> * one of the main things is to help another person test easily so you have to assume they know nothing.
>   e.g of such a manual test
> 
> 1. Start the server `npm run start`
> 2. Post the sample data below to the endpoint `/api/v1/articles`
> 
> ```json
> {
>   "name": "Yes",
>   "message": "No"
> }
> ```
> 
> 3. It should successfully return `so so and so` response
> 
> * Screenshots will not be too necessary in that case
> * Also link the relevant issue if possible

I have learned a lot this past 15 hours. I will make those changes going forward. Thanks so much

**Comment by jocrah** [comment]
> Sorry for the late reply, i reached out to the backend team on this and apparently it was agreed that we change the names of certain properties.

no problem. you're not obliged to reply instantly in this case. Reply when you're able to

> On that note, i wouldn't have to change mine then. Meaning my PR isn't a WIP anymore.

Cool, understood. Thanks for the update

**Comment by jocrah** [comment]
@beblicarl, kindly let me know when you are done with the changes so we wrap up the review 🙂

**Comment by jocrah** [comment]
> @beblicarl, kindly let me know when you are done with the changes so we wrap up the review 🙂

no pressure, just saying this in case you might want me to review and I don't know

**Comment by beblicarl** [comment]
> > @beblicarl, kindly let me know when you are done with the changes so we wrap up the review slightly_smiling_face
> 
> no pressure, just saying this in case you might want me to review and I don't know

I am done now, kindly review the changes for me

**Comment by beblicarl** [comment]
> @beblicarl, kindly let me know when you are done with the changes so we wrap up the review slightly_smiling_face

I am done now, kindly review the changes for me

File: src/routes/users.js [hunk 3]
```diff
@@ -3,28 +3,24 @@ const db = require('../db')
...
-.get(fetchUsers)
-.post(createUsers)
+router.route('/').get(fetchUsers)
+router.route('/').post(createUsers)
 
-router.route('/:id')
-.get(getUser)
-.patch(updateUser)
-.delete(deleteUser)
+router.route('/:id').get(getUser)
```
Comment (jocrah) [inline src/routes/users.js:22]: @beblicarl why this change from how it was?

Comment (beblicarl) [reply src/routes/users.js:22]: you commented on how it was all on one line so I assumed you wanted it to be multilined @jocrah

Comment (jocrah) [reply src/routes/users.js:22]: haha okay got it.

i only asked a question to understand why the change was made. I think what was discussed was to reset the changes to even exclude this file because if it should be multi-lined, what was already there before the pr had covered that. This is a bit different and more verbose

Comment (jocrah) [reply src/routes/users.js:22]: yea and I understand the assumptions but just a heads up that my questions don't really mean you should perform an action or make a change, it's just a question. If there's an explicit change, I will let you know

Comment (beblicarl) [reply src/routes/users.js:22]: > haha okay got it.
> 
> i only asked a question to understand why the change was made. I think what was discussed was to reset the changes to even exclude this file because if it should be multi-lined, what was already there before the pr had covered that. This is a bit different and more verbose

okay, I will reset it then

Comment (beblicarl) [reply src/routes/users.js:22]: > yea and I understand the assumption but just a heads up that my questions don't really mean you should perform an action or make a change, it's just a question. If there's an explicit change, I will let you know

okay

**Comment by jocrah** [comment]
also not seeing a way of manual testing yet from your description. Screenshots are great for UI work but backend will not really yield much. Could you help with the kind of data you're using to test.

Like the example i gave in my earlier comment? https://github.com/tbf-challenge/teamwork-be/pull/56#issuecomment-1216336553

**Comment by beblicarl** [comment]
> also not seeing a way of manual testing yet from your description. Screenshots are great for UI work but backend will not really yield much. Could you help with the kind of data you're using to test.
> 
> Like the example i gave in my earlier comment? [#56 (comment)](https://github.com/tbf-challenge/teamwork-be/pull/56#issuecomment-1216336553)

I have described how all 4 of my endpoints should be manually tested, I am not sure why it does not appear up there @jocrah 

Screenshot from 2022-08-16 17-36-19

File: src/routes/users.js (hunk 3 above)
Comment (beblicarl) [reply src/routes/users.js]: > haha okay got it.
> 
> i only asked a question to understand why the change was made. I think what was discussed was to reset the changes to even exclude this file because if it should be multi-lined, what was already there before the pr had covered that. This is a bit different and more verbose

with regards to this, I have reset it and pushed it to the target branches now

File: src/routes/users.js (hunk 2 above)
Comment (beblicarl) [reply src/routes/users.js]: @jocrah  i have reset the user.js now

**Comment by beblicarl** [comment]
> also not seeing a way of manual testing yet from your description. Screenshots are great for UI work but backend will not really yield much. Could you help with the kind of data you're using to test.
> 
> Like the example i gave in my earlier comment? [#56 (comment)](https://github.com/tbf-challenge/teamwork-be/pull/56#issuecomment-1216336553)

I have worked on the pre-submission checklist now @jocrah

File: src/app.js [hunk 4]
```diff
@@ -3,18 +3,72 @@ const morgan = require('morgan')
...
+if (process.env.NODE_ENV === 'development') {
     app.use(morgan('dev'))
 }
 app.get('/', (req, res) => {
     res.send('Hello friend')
 })
 
+// CREATE USER ENDPOINT
+
+app.post('/users', async (req, res) => {
```
Comment (jocrah) [inline src/app.js:21]: @beblicarl please is this within the scope or a personal addition? I was thinking this was being worked on in another story here https://github.com/tbf-challenge/teamwork-be/pull/65/files

File: src/routes/posts.js [hunk 5]
```diff
@@ -0,0 +1,26 @@
...
+    try {
+        const { id, userId, title, image, content, published, createdAt } =
+            req.body
+        const newArticle = await db.query(
+            'INSERT INTO posts (id , "userId", title , image , content , published , "createdAt" ) VALUES ($1 , $2 , $3 , $4 , $5 ,$6 , $7) RETURNING *',
+            [id, userId, title, image, content, published, createdAt]
+        )
+        res.json(newArticle)
+    } catch (err) {
+        console.error(err.message)
```
Comment (jocrah) [inline src/routes/posts.js:19]: @beblicarl  we would want to know the error from the endpoint itself 

e.g frontend can't get access to this console.log to be able to know there's been an error 

make line 9, `async (req, res, next)` and use pass the error to `next` like so

```suggestion
  next(err)
```

PS: You can still keep the logger for your own purposes before the `next(err)` anyway

File: src/routes/posts.js [hunk 6]
```diff
@@ -0,0 +1,26 @@
+const express = require('express')
+
+const db = require('../db')
+
+const router = express.Router()
+
+// CREATE AN ARTICLE
+
+const createPosts = async (req, res) => {
```
Comment (jocrah) [inline src/routes/posts.js:9]: i think `createPost` will cover it if its just one we're creating? so its easier for a new person to read and understand

File: src/app.js [hunk 7]
```diff
@@ -3,18 +3,72 @@ const morgan = require('morgan')
...
+        )
+        res.json(newUser)
+    } catch (err) {
+        console.error(err.message)
+    }
+})
+
+//  CREATE POST ENDPOINT
+
+app.post('/article', async (req, res) => {
```
Comment (jocrah) [inline src/app.js:58]: i don't want to focus on structure too much for now till necessary but I think this should be handled in the `posts.js`.  

The code flow for now is a bit confusing for me so lets do that and see how it goes

**Comment by jocrah** [comment]
@beblicarl I think we can merge after these changes if all checks out. Kindly let me know if any part isn't clear. Thanks

File: src/app.js (hunk 7 above)
Comment (beblicarl) [reply src/app.js:58]: Done

File: src/app.js (hunk 4 above)
Comment (beblicarl) [reply src/app.js:21]: Yes, that's true

File: src/routes/posts.js (hunk 6 above)
Comment (beblicarl) [reply src/routes/posts.js]: I have worked on this review now

File: src/routes/posts.js (hunk 5 above)
Comment (beblicarl) [reply src/routes/posts.js:29]: i have worked on this review too, I included the console.error for my own purposes

File: src/routes/posts.js (hunk 6 above)
Comment (jocrah) [reply src/routes/posts.js]: okay, yet to push?

File: src/routes/posts.js [hunk 8]
```diff
@@ -0,0 +1,37 @@
...
+const dateTime = new Date().toJSON()
+
+// CREATE AN ARTICLE
+
+const createPosts = async (req, res, next) => {
+    try {
+        const { id, userId, title, image, content, published, createdAt } =
+            req.body
+        await db.query(
+            'INSERT INTO posts (id , "userId", title , image , content , published , "createdAt" ) VALUES ($1 , $2 , $3 , $4 , $5 ,$6 , $7) RETURNING *',
```
Comment (jocrah) [inline src/routes/posts.js:16]: @beblicarl, from the postgres setup migration added, the `createdAt` should be populated by default when the document is inserted.

https://github.com/tbf-challenge/teamwork-be/blob/b4a4f5817504c2590f509112e27514c240e2fc14/migrations/1660111562812_db-init.js#L9 

That means you will use the result after the insert to get the fields and then return that as the API response

Not tested at my end so can tell the response that comes from the insert but hopefully it returns the inserted document

As always, ask me anything that you want clarification on

File: src/routes/posts.js (hunk 6 above)
Comment (beblicarl) [reply src/routes/posts.js]: Screenshot from 2022-08-17 12-30-07
 I have pushed it already, 30 mins ago, please confirm if you see it now

Comment (jocrah) [reply src/routes/posts.js]: @beblicarl yea got it. this change was just about the naming of the function. since its one post being created at a go. just `createPost` without the `s` is fine

File: src/routes/posts.js (hunk 8 above)
Comment (jocrah) [reply src/routes/posts.js:16]: so in effect, what i'm saying is you don't have to manually insert the `createdAt` yourself. Let the database do that

Comment (jocrah) [reply src/routes/posts.js:16]: @beblicarl just saw we're also inserting the id field from the request. that shouldn't be the case

 yea so same for the `id` field especially. It should be assigned from the database and not passed on from the frontend or anyone calling the endpoint

**Comment by jocrah** [comment]
mistake, apologies

**Comment by beblicarl** [comment]
> things are looking good so far @beblicarl, great work. maybe there'll be some other refactors here and there as the code size grows but we're getting to a cool level on this one.

thank you

File: src/routes/posts.js (hunk 8 above)
Comment (beblicarl) [reply src/routes/posts.js:16]: > @beblicarl, from the postgres setup migration added, the `createdAt` should be populated by default when the document is inserted.
> 
> https://github.com/tbf-challenge/teamwork-be/blob/b4a4f5817504c2590f509112e27514c240e2fc14/migrations/1660111562812_db-init.js#L9
> 
> That means you will use the result after the insert to get the fields and then return that as the API response
> 
> Not tested at my end so can 't ell the response that comes from the insert but hopefully it returns the inserted document
> 
> As always, ask me anything that you want clarification on

okay, I will work on this as well

Comment (beblicarl) [reply src/routes/posts.js:16]: > so in effect, what i'm saying is you don't have to manually insert the `createdAt` yourself. Let the database do that

I will work on this as well

Comment (beblicarl) [reply src/routes/posts.js:14]: I am done with both the id and createdAt properties @jocrah

Comment (jocrah) [reply src/routes/posts.js:14]: okay cool, taking a look

File: src/routes/posts.js [hunk 9]
```diff
@@ -0,0 +1,35 @@
...
+
+const db = require('../db')
+
+const router = express.Router()
+
+// CREATE AN ARTICLE
+
+const createPosts = async (req, res, next) => {
+    try {
+        const { id, userId, title, image, content, published, createdAt } =
```
Comment (jocrah) [inline src/routes/posts.js:11]: @beblicarl not sure if maybe there are more changes yet to be done but just to clarify it seems we're still getting it from the`req.body` though and inserting it into the db. these are autogenerated fields by the database itself so should not be gotten from the request

 we should insert without the id and createdAt and after the insert use the response from the db to get the id and created at

File: src/routes/posts.js (hunk 6 above)
Comment (beblicarl) [reply src/routes/posts.js]: I have just seen this, i will work on it as well

File: src/routes/posts.js (hunk 9 above)
Comment (beblicarl) [reply src/routes/posts.js:11]: alright

Comment (beblicarl) [reply src/routes/posts.js:11]: also, how do i send id and createdAt in the response body if I don't get it from the request body ? @jocrah 
image

File: src/routes/posts.js [hunk 10]
```diff
@@ -0,0 +1,35 @@
...
+
+const router = express.Router()
+
+// CREATE AN ARTICLE
+
+const createPost = async (req, res, next) => {
+    try {
+        const { id, userId, title, image, content, published, createdAt } =
+            req.body
+        await db.query(
```
Comment (jocrah) [inline src/routes/posts.js:13]: > also, how do i send id and createdAt in the response body if I don't get it from the request body ? @jocrah

```suggestion
       const newPost = await db.query(
```

@beblicarl, by assigning the insert to a variable and using that. Don't really know the response that will be in the variable but ideally the info should be there

Comment (beblicarl) [reply src/routes/posts.js:13]: so it would be newPost.createdAt and newPost.id ?

Comment (jocrah) [reply src/routes/posts.js:13]: correct.  kindly do check what the command returns and lets see if it has the data. you can share it here if possible

Comment (beblicarl) [reply src/routes/posts.js:13]: image

I am getting this error, newArticle is never reassigned

Comment (jocrah) [reply src/routes/posts.js:13]: cool cool, making progress. it means we need to do a bit more work then.

the `id` column type must be updated in the migrations to autogenerate on its own in the database then.

for sure we can't be taking ids from the request for a primary key field because it means the frontend always has to know what key to send that will not conflict with the existing keys in the db

found this so far so do take a look at it and let me know any ideas you come up with: https://chartio.com/resources/tutorials/how-to-define-an-auto-increment-primary-key-in-postgresql/

what we want for now is to create a new migration and modify the id column so that it autogenerates on its own. I'll join you on that later but do let me know what you find :wink:

Comment (beblicarl) [reply src/routes/posts.js:13]: that's perfect !
How about the createdAt as well ?

Comment (beblicarl) [reply src/routes/posts.js:13]: i don't understand what you mean with the current_timestamp

Comment (jocrah) [reply src/routes/posts.js:13]: cool. glad you're asking questions :smile:. so i meant the existing migration should handle that, if you see line 9. So you can just try passing everything else and exclude the `createdAt` and see if the insert works

https://github.com/tbf-challenge/teamwork-be/blob/b4a4f5817504c2590f509112e27514c240e2fc14/migrations/1660111562812_db-init.js#L6-L10

Comment (jocrah) [reply src/routes/posts.js:13]: for more context, the `current_timestamp` is a key word in most SQL definitions to say it should use the date and time at that particular instant as the default.

https://www.postgresqltutorial.com/postgresql-date-functions/postgresql-current_timestamp/

Comment (beblicarl) [reply src/routes/posts.js:13]: oh okay, i carefully read through it.. so how i understand it like you said early on, the id as long as the current_timestamp should be handled in the migration

Comment (jocrah) [reply src/routes/posts.js:13]: maybe we're saying the same thing but i mean for the `createdAt`, the default being `current_timestamp` means it will always be generated correctly for you if you insert a document. You don't have to create your own date and pass it along

Comment (beblicarl) [reply src/routes/posts.js:13]: yeah, i meant the same thing. What I mean is, during the creation of the various tables ;
1.  id takes a value of a SERIAL PRIMARY key, which means it will increment automatically anytime a new document is added
2. createdAt has a default value being current_timestamp, which means it automatically provides the current time anytime a document is added.


Just to be sure, are we on the same page?

Comment (jocrah) [reply src/routes/posts.js:13]: Exactly, yes we are on the same page

Comment (beblicarl) [reply src/routes/posts.js:13]: awesome. I guess i'd wait till later for you to join me with the other things

Comment (jocrah) [reply src/routes/posts.js:13]: @beblicarl seems @jakazzy was also experiencing an id issue and has worked on the [fix](https://github.com/tbf-challenge/teamwork-be/pull/75) for that. so will merge that so you can pull those changes into yours and see how it goes.

You can ask her the reason for the changes she made to learn more. Kindly follow the manual steps in the pr so everything works smoothly

After that. we can go again on this and see how it works out

**Comment by beblicarl** [comment]
I have done everything you asked me to do, kindly review it for me @jocrah

**Review by jocrah — APPROVED** [review]

**Comment by jocrah** [comment]
> I have done everything you asked me to do, kindly review it for me @jocrah

yea manual test works at my end. merging now :rocket:

**Comment by beblicarl** [comment]
Thanks so much @jocrah . I will apply similar changes on the other PRs as well so you can review them

**Comment by jocrah** [comment]
> Thanks so much @jocrah . I will apply similar changes on the other PRs as well so you can review them

Great, let me know when that’s done
